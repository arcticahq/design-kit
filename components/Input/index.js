import React from "react"
import PropTypes from "prop-types"
import classnames from "classnames"
import { withFormsy, propTypes } from "formsy-react"

class BaseInput extends React.Component {
  state = { blurred: false }

  render() {
    const {
      isValid,
      isFormSubmitted,
      getErrorMessage,
      getValue,
      setValue,
      name,
      title,
      type,
      required,
      styles
    } = this.props
    const { blurred } = this.state

    const errorMessage =
      (blurred && !isValid()) || (!isValid() && isFormSubmitted())

    const inputClass = classnames(styles.input, {
      [styles.input]: errorMessage
    })
    const labelClass = classnames(styles.labelClass, {
      [styles.labelError]: errorMessage
    })
    const { errorClass } = styles

    return (
      <>
        <label className={labelClass} htmlFor={name}>
          {title}
          {!required && <span className={styles.optional}> (optional)</span>}
          <input
            aria-required={required}
            aria-invalid={!!errorMessage}
            onFocus={() => this.setState({ blurred: false })}
            onBlur={() => this.setState({ blurred: true })}
            onChange={e => setValue(e.target.value)}
            value={getValue() || ``}
            name={name}
            className={inputClass}
            required={required}
            id={name}
            type={type}
          />
        </label>
        {errorMessage && (
          <p role="alert" className={errorClass}>
            {getErrorMessage()}
          </p>
        )}
      </>
    )
  }
}

BaseInput.propTypes = {
  ...propTypes,
  type: PropTypes.string,
  styles: PropTypes.object
}

BaseInput.defaultProps = {
  type: `text`,
  styles: {
    input:
      "appearance-none mt-2 mb-6 block w-full bg-gray-100 text-gray-darker border py-3 px-3 mb-3 leading-tight focus:outline-none focus:bg-white focus:border-indigo-400",
    inputError: "border-red-600",
    label: "block uppercase tracking-wide text-black text-xs font-bold",
    labelError: "text-red-600",
    error: "text-red-600 text-xs italic absolute -mt-6",
    optional: "text-gray-600"
  }
}

export default BaseInput
export const FormsyInput = withFormsy(BaseInput)
