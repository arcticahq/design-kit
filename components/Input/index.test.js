import React from "react"
import { mount } from "enzyme"
import Input from "."

describe(`Input`, () => {
  let props
  let mountedComponent
  const wrapper = () => {
    if (!mountedComponent) {
      mountedComponent = mount(<Input {...props} />)
    }
    return mountedComponent
  }

  beforeEach(() => {
    props = {
      name: `name`,
      title: `Your name`,
      type: `text`,
      required: true,
      showRequired: () => true,
      isPristine: () => true,
      isValid: jest.fn(),
      isFormSubmitted: jest.fn(),
      getValue: jest.fn(),
      setValue: jest.fn(),
      showError: jest.fn(),
      getErrorMessage: jest.fn()
    }
    mountedComponent = undefined
  })

  it(`renders an Input correctly`, () => {
    expect(wrapper()).toMatchSnapshot()
  })

  it("Checks for errors when the input is blurred", () => {
    wrapper()
      .find("input")
      .simulate("focus")
      .simulate("blur")
    expect(wrapper().state("blurred")).toEqual(true)
    expect(props.getErrorMessage).toBeCalled()
  })

  // shows an error when invalid

  // clears errror when valid

  it("Calls setValue() when the input is changed", () => {
    wrapper()
      .find("input")
      .simulate("change", { target: { value: "My name" } })
    expect(props.setValue).toBeCalled()
  })

  // does not contain required label or text when not required
})
