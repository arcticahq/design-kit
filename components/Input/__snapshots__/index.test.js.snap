// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders an Input correctly 1`] = `
<Input
  getErrorMessage={[MockFunction]}
  getValue={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isFormSubmitted={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  isPristine={[Function]}
  isValid={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  name="name"
  required={true}
  setValue={[MockFunction]}
  showError={[MockFunction]}
  showRequired={[Function]}
  title="Your name"
  type="text"
>
  <label
    className="block uppercase tracking-wide text-black text-xs font-bold"
    htmlFor="name"
  >
    Your name
    <input
      aria-invalid={false}
      aria-required={true}
      className="appearance-none mt-2 mb-6 block w-full bg-gray-100 text-gray-900 border rounded-lg py-3 px-3 mb-3 leading-tight focus:outline-none focus:bg-white focus:border-yellow-500"
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="text"
      value=""
    />
  </label>
</Input>
`;
